### PR TITLE
#26/ Store messages in in-memory BadgerDB

### PR DESCRIPTION
**Pull request by DarjaLeonova**

### CODE REVIEW

File: repo/repo.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+func NewRepo(db *badger.DB) *Repo {
+	return &Repo{
+		db: db,
+	}
+}
+
+func Connect() (*badger.DB, error) {
+	db, err := badger.Open(badger.DefaultOptions("").WithInMemory(true))
+	if err != nil {
+		log.Fatal(err)
```
Comment (janishorsts) [inline repo/repo.go]:
```suggestion
		return nil, fmt.Errorf("connect to DB: %w", err)
```

File: pkg/translate/translate.go [hunk 2]
```diff
@@ -65,7 +68,12 @@ func (t *TranslateServiceServer) UploadTranslationFile(
 	// Save to DB/FS...
 	messages.Language = language
 	messages.Labels = reqLabels
-	_ = messages
+	msg := messages
+
+	err = t.repo.SaveMessages(msg)
```
Comment (janishorsts) [inline pkg/translate/translate.go]:
```suggestion

	err = t.repo.SaveMessages(messages)
```

File: repo/repo.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	return db, nil
+}
+
+func (repo *Repo) SaveMessages(m model.Messages) error {
+	for _, value := range m.Messages {
```
Comment (janishorsts) [inline repo/repo.go]: `Messages` is a resource to be saved. Serialize it to JSON. The key is translation id.

**Review by janishorsts — CHANGES_REQUESTED** [review]

File: pkg/translate/translate.go [hunk 4]
```diff
@@ -65,7 +68,12 @@ func (t *TranslateServiceServer) UploadTranslationFile(
 	// Save to DB/FS...
```
Comment (VladislavsPerkanuks) [inline pkg/translate/translate.go]: You can delete this comment now, as you implemented `save`

File: cmd/translate/main.go [hunk 5]
```diff
@@ -22,19 +26,29 @@ type TranslateServiceServer struct {
 const serverAddr = "localhost:8080"
 
 func main() {
+	db, err := repo.Connect()
```
Comment (VladislavsPerkanuks) [inline cmd/translate/main.go:29]: No error check

File: pkg/translate/translate.go [hunk 6]
```diff
@@ -65,7 +68,12 @@ func (t *TranslateServiceServer) UploadTranslationFile(
...
 	messages.Language = language
 	messages.Labels = reqLabels
-	_ = messages
+	msg := messages
+
+	err = t.repo.SaveMessages(msg)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "saving messages: %w", err)
```
Comment (VladislavsPerkanuks) [inline pkg/translate/translate.go]: `codes.Internal`

File: pkg/repo/repo.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = repo.db.Update(func(txn *badger.Txn) error {
+		err = txn.Set([]byte(id), messagesJson)
+		if err != nil {
+			err = fmt.Errorf("setting key/value pairs %w", err)
+		}
+		return err
```
Comment (janishorsts) [inline pkg/repo/repo.go:41]:
```suggestion
		return nil
```

File: pkg/repo/repo.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+func (repo *Repo) SaveMessages(id string, m model.Messages) error {
+	messagesJson, err := encodeMessages(m)
+	if err != nil {
+		return err
+	}
+
+	err = repo.db.Update(func(txn *badger.Txn) error {
+		err = txn.Set([]byte(id), messagesJson)
+		if err != nil {
+			err = fmt.Errorf("setting key/value pairs %w", err)
```
Comment (janishorsts) [inline pkg/repo/repo.go:39]:
```suggestion
			return fmt.Errorf("setting key/value pairs %w", err)
```

File: pkg/repo/repo.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+func encodeMessages(m model.Messages) ([]byte, error) {
+	data, err := json.Marshal(m)
+	if err != nil {
+		err = fmt.Errorf("model.Messages marshaliing to JSON %w", err)
+		return nil, err
+	}
+
+	return data, err
+}
```
Comment (VladislavsPerkanuks) [inline pkg/repo/repo.go]: Redundant function

Comment (DarjaLeonova) [reply pkg/repo/repo.go]: So you propose to do everything in single function?

Comment (VladislavsPerkanuks) [reply pkg/repo/repo.go]: yeah, encode messages does one thing which is marshal
so: `data,err := json.Marshal(m) == data,err := json.encodeMessages(m)`

Comment (DarjaLeonova) [reply pkg/repo/repo.go]: done ;)

File: pkg/repo/repo.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	return db, nil
+}
+
+func (repo *Repo) SaveMessages(id string, m model.Messages) error {
+	messagesJson, err := json.Marshal(m)
+	if err != nil {
+		err = fmt.Errorf("marshaling model.Messages to JSON")
+		return err
```
Comment (VladislavsPerkanuks) [inline pkg/repo/repo.go]:
```suggestion
return fmt.Errorf("marshaling model.Messages to JSON: %w", err)
```

File: pkg/repo/repo.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, fmt.Errorf("connect to DB: %w", err)
+	}
+
+	return db, nil
+}
+
+func (repo *Repo) SaveMessages(id string, m model.Messages) error {
+	messagesJson, err := json.Marshal(m)
+	if err != nil {
+		return fmt.Errorf("marshaling model.Messages to JSON")
```
Comment (VladislavsPerkanuks) [inline pkg/repo/repo.go]:
```suggestion
		return fmt.Errorf("marshaling model.Messages to JSON: %w", err)
```
😅

Comment (DarjaLeonova) [reply pkg/repo/repo.go]: AAAAA

Comment (DarjaLeonova) [reply pkg/repo/repo.go]: I will wait for more comments ;D

Comment (VladislavsPerkanuks) [reply pkg/repo/repo.go]: No more comments from me :)

**Review by VladislavsPerkanuks — APPROVED** [review]

File: pkg/repo/repo.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func encodeMessages(m model.Messages) ([]byte, error) {
+	data, err := json.Marshal(m)
+	if err != nil {
+		err = fmt.Errorf("model.Messages marshaliing to JSON %w", err)
+		return nil, err
+	}
+
+	return data, err
```
Comment (janishorsts) [inline pkg/repo/repo.go]:
```suggestion
	return data, nil
```

**Review by janishorsts — CHANGES_REQUESTED** [review]

**Pull request merged**